### ISSUE TITLE
MMS Restore

### ISSUE THREAD

**Issue by meliodas926** [jberkel/sms-backup-plus#784]
I can't find a way to access my backed up MMS messages.  I can't restore them either since the feature is not yet available.  Is there any way at all to get the pictures?  I need some very bad!
Thank you

SMS backup+

**Comment by nathan323** [comment]
They will all be on your Gmail account.

When will this feature be supported? If it's a long conversation history, missing many MMS messages defeats the purpose of a backup. Is there work in progress?

**Comment by jberkel** [comment]
It's not planned: https://github.com/jberkel/sms-backup-plus/#faq-restore-MMS

**Comment by b8two** [comment]
I agree that MMS messages are an important part of a messaging thread. 
I have a suggestion as an alternative to supporting MMS restore that we might all be happy with.

Instead of creating a MMS on the device, create a SMS with the Text (from the MMS) and a Link to the Media (that's added as an attachment in the email backup). That way the messaging thread is kept and the content is view able just not within the SMS app. There would need to be a Special TAG in the SMS to indicated it was a restored MMS and not needing to be backed up uniquely.
If you can't link to the email attachment, a link to the email would do (if possible to load up in gmail?).
If that is too difficult, perhaps the attachments can be stored on Google Drive and linked in that way. 
Could use a URL shorter and I was going to suggest https://goo.gl/  but I have just read that google are going to shut this down and replace it with Firebase Dynamic Links which looks like a better option for this usage case. https://firebase.google.com/products/dynamic-links/
i.e. the attachment could be stored in firebase (linked account) with 5 GB of free storage.

**Comment by jberkel** [comment]
@b8two interesting idea! You can link to specific emails in Gmail, (e.g. `https://mail.google.com/mail/u/0/#inox/[1636afe1d0f09b90]`) but I'm not sure if these URLs are stable, and I couldn't find any documentation (how to determine the id part of the url). If you can find anything, let me know.

In general I'm hesitant to add dependencies to other (non-free) services such as Drive or Firebase.

**Comment by b8two** [comment]
Hi,

I just played with this online API function to get a list of ID's (Appears to be stable if it's part of the API);
https://developers.google.com/gmail/api/v1/reference/users/messages/list

Then used the ID to get the related message;
https://developers.google.com/gmail/api/v1/reference/users/messages/get

& hence I could use this ID with the URL (without the square brackets); https://mail.google.com/mail/u/0/#inbox/1636afe1d0f09b90

BTW, Google Drive is 15GB Free and is shared with the gmail inbox Quota not really a separate service.

**Comment by b8two** [comment]
just to re-assure you about the ID.

id | string | The immutable ID of the message.

From: https://developers.google.com/gmail/api/v1/reference/users/messages

**Comment by b8two** [comment]
I guess this would also be a way to check for uniqueness of the SMS/MMS regarding avoiding duplication backups. If message contains; "https://mail.google.com/mail/u/0/#inbox/1636afe1d0f09b90" and Message ID "1636afe1d0f09b90" exists, skip

**Comment by jberkel** [comment]
thanks for doing the research. Do the links open correctly  in the Gmail app on the phone?

I need to get the message ID from the IMAP conversation, which seems to be possible with an extension:

> Access to the Gmail unique message ID: X-GM-MSGID
>
> Gmail provides a unique message ID for each email so that a unique message may be identified across multiple folders. Retrieval of this message ID is supported via the X-GM-MSGID attribute on the FETCH command. The message ID is a 64-bit unsigned integer and is the decimal equivalent for the ID hex string used in the web interface and the Gmail API.

https://developers.google.com/gmail/imap/imap-extensions

**Comment by b8two** [comment]
I have tested but I was unable to find the right link to open on the phone.  The links work well on a desktop web browser but trying on a mobile, gmail tries re-directs to a mobile version of the page, dropping the message thread ID.   if you replace the #inbox with #all, then any message ID will link correctly on a desktop.  

I could not find any threads about links that work on a mobile phone.  just ways to view messages in Original View to get different ID's etc. It appears that it would make more sense to put a copy of the attachment on cloud storage and put a link to that.
i.e.  https://drive.google.com/file/d/0B7RLMHD6ozpSMEZmRUhPMXFOMjg4Q1JDMjZfRWdaZTltX2xj/view
^ this file was added to google drive via gmail with the option "Save to Drive" (Icon in Gmail next to download),

-  I was able to move the file (on google drive) to a "MMS" sub folder (but it keeps the same link as above).
- It's still google storage (same Auth) 
- The storage isn't public.
- When I click the link in a SMS, Google Drive app opens the link and (probably because I have two gmail accounts on my phone) checks which account to open it from (authenticated)

I understand that you don't have access to do this via IMAP interface but perhaps this is another check box option in the app, 
Backup "MMS attachment on Drive Storage". 
Restore "MMS attachements linked to Drive Storage"

**Comment by InfOracle** [comment]
I have a Note 8 on Verizon and I see my MMS messages in Gmail, and SOME did come back after a Restore, but not all...
